## Contents

1. [Importing Libraries](#1.-Importing-Libraries)     
	1.1. [Importing Dataframes](#1.1-Importing-Dataframes)      
2. [Aggregate the mean of the order_number column grouped by department id for the entire dataframe](#2.-Aggregate-the-mean-of-the-order-number-column-grouped-by-department-id-for-the-entire-dataframe)   
3. [Analysing the difference between the means of the two dataframes](#3.-Analysing-the-difference-between-the-means-of-the-two-dataframes)   
4. [Creating a loyalty flag for existing customers using the transform() and loc() functions.](#4.-Creating-a-loyalty-flag-for-existing-customers-using-the-transform()-and-loc()-functions.)   
5. [Is there a difference between the spending habits of the three types of customers you identified?](#5.-Is-there-a-difference-between-the-spending-habits-of-the-three-types-of-customers-you-identified?)   
6. [Creating a spending flag for each user based on the average price across all their orders](#6.-Creating-a-spending-flag-for-each-user-based-on-the-average-price-across-all-their-orders)    
7. [Creating an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column to determine frequent versus non-frequent customers](#7.-Creating-an-order-frequency-flag-that-marks-the-regularity-of-a-user’s-ordering-behavior-according-to-the-median-in-the-“days-since-prior-order”-column-to-determine-frequent-versus-non-frequent-customers)    
8. [Exporting the dataframe as a pickle file](#8.-Exporting-the-dataframe-as-a-pickle-file)    

## 1. Importing Libraries

In [1]:
import pandas as pd
import numpy as np
import os

### 1.2. Importing Dataframes

In [2]:
# Folder path
path = r'C:\Users\User 1\Documents\Instacart Basket Analysis 04-2023'

In [3]:
path

'C:\\Users\\User 1\\Documents\\Instacart Basket Analysis 04-2023'

In [4]:
#Importing 'orders_products_merged_variables.pkl'
df_ords_prods_merged = pd.read_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged_variables.pkl'))

## 2. Aggregate the mean of the order_number column grouped by department id for the entire dataframe

In [5]:
# Grouping the data by 'department_id' and then aggregating the data to find the mean of 'order_number'
df_ords_prods_merged.groupby('department_id')['order_number'].mean()

department_id
1     15.457838
2     17.277920
3     17.170395
4     17.811403
5     15.215751
6     16.439806
7     17.225802
8     15.340650
9     15.895474
10    20.197148
11    16.170638
12    15.887671
13    16.583536
14    16.773669
15    16.165037
16    17.665606
17    15.694469
18    19.310397
19    17.177343
20    16.473447
21    22.902379
Name: order_number, dtype: float64

## 3. Analysing the difference between the means of the two dataframes

The results for the entire dataframe contain the full list of 'department_id whereas the subset is capped to 8. 
The mean calculated from the entire dataframe are generally lower than the mean calculated from the subset. 
For example, the mean for department_id: 4 is 17.8 in the entire df whereas the mean for department_id: 4 in the subset is 18.8. 
However, there is an exception: Department_id 17 has a higher mean value (15.7) compared to its subset (11.3).

## 4. Creating a loyalty flag for existing customers using the transform() and loc() functions.

In [6]:
#Creating a new column containing the maximum frequency of the 'order_number' column
df_ords_prods_merged['max_order'] = df_ords_prods_merged.groupby(['user_id'])['order_number'].transform(np.max)

In [7]:
# Checking the max_ order column
df_ords_prods_merged.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  aisle_id  department_id  prices _merge    price_range_loc  \
0          Soda        77              7     9.0   both  Mid-range product   
1          Soda        77              7     9.0   both  Mid-range product   
2          Soda        77              7     9.0   both  Mid-range product   
3          Soda        77              7     9.0   both  Mid-range product   
4          Soda        77              7     9.0   both  Mid-range product   
5          Soda        77              7     9.0   both  Mid-range product   
6          Soda        77              7     9.0   both  Mid-range product   
7          Soda        77              7     9.0   both  Mid-range product   
8          Soda        77              7     9.0   both  Mid-range product   
9          Soda        77              7     9.0   both  Mid-range product   
10         Soda        77              7     9.0   both  Mid-range product   
11         Soda        77              7     9.0   both  Mid-range product   
12         Soda        77              7     9.0   both  Mid-range product   
13         Soda        77              7     9.0   both  Mid-range product   
14         Soda        77              7     9.0   both  Mid-range product   

       busiest_day  Busiest_days busiest_period_of_day  max_order  
0   Regularly busy  Regular days           Most orders         10  
1   Regularly busy  Slowest days        Average orders         10  
2   Regularly busy  Slowest days           Most orders         10  
3       Least busy  Slowest days        Average orders         10  
4       Least

In [8]:
# Creating a flag that assigns a loyalty level to a user_id based on their corresponding max order value.
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [9]:
df_ords_prods_merged.loc[(df_ords_prods_merged['max_order'] <= 40) & (df_ords_prods_merged['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [10]:
df_ords_prods_merged.loc[df_ords_prods_merged['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [11]:
#Checking the accuracy and count of 'loyalty_flag' values and labels
df_ords_prods_merged['loyalty_flag'].value_counts(dropna = False)

Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: loyalty_flag, dtype: int64

In [12]:
df_ords_prods_merged[['user_id', 'loyalty_flag', 'order_number']].head(60)

user_id      loyalty_flag  order_number
0         1      New customer             1
1         1      New customer             2
2         1      New customer             3
3         1      New customer             4
4         1      New customer             5
5         1      New customer             6
6         1      New customer             7
7         1      New customer             8
8         1      New customer             9
9         1      New customer            10
10       15  Regular customer            15
11       15  Regular customer            17
12       15  Regular customer            18
13       15  Regular customer            21
14       15  Regular customer            22
15       19      New customer             2
16       19      New customer             5
17       19      New customer             7
18       21  Regular customer            10
19       31  Regular customer            10
20       31  Regular customer            17
21       43  Regular customer             6
22       43  Regular customer             9
23       52  Regular customer             1
24       52  Regular customer             2
25       52  Regular customer             3
26       52  Regular customer             5
27       52  Regular customer             7
28       52  Regular customer             9
29       52  Regular customer            10
30       52  Regular customer            11
31       52  Regular customer            12
32       52  Regular customer            15
33       52  Regular customer            17
34       52  Regular customer            18
35       52  Regular customer            19
36       52  Regular customer            21
37       67  Regular customer             1
38       67  Regular customer             2
39       67  Regular customer             3
40       67  Regular customer             4
41       67  Regular customer             5
42       67  Regular customer             6
43       67  Regular customer             8
44       67  Regular customer             9
45       67  Regular customer            10
46       67  Regular customer            11
47       67  Regular customer            12
48       67  Regular customer            13
49       67  Regular customer            14
50       67  Regular customer            16
51       67  Regular customer            18
52       67  Regular customer            19
53       67  Regular customer            21
54       67  Regular customer            22
55       67  Regular customer            23
56       81      New customer             3
57       81      New customer             5
58       82  Regular customer             1
59       82  Regular customer             3

In [13]:
df_ords_prods_merged.shape

(32404859, 20)

## 5. Is there a difference between the spending habits of the three types of customers you identified?

In [14]:
# Checking the descriptive statistics of the three types of customers by performing multiple aggregations ('min', 'max', 'mean', 'sum')
df_ords_prods_merged.groupby('loyalty_flag').agg({'prices': ['min', 'max', 'mean', 'sum']})

prices                              
                    min   max      mean           sum
loyalty_flag                                         
Loyal customer      1.0  25.0  7.772831  7.993651e+07
New customer        1.0  25.0  7.800320  4.870512e+07
Regular customer    1.0  25.0  7.797431  1.237981e+08

The mean for New customers are slightly higher than 'Loyal' and 'Regular' customers. This means that on average, new customers purchase higher priced products than 'Loyal' and 'Regular' customers.
There is no difference in the minimum and maximum prices suggesting that all types of customers have purchased the cheapest products as well as the priciest of products.
The sum of prices is highest with 'Loyal customers'. This means that loyal customers purchase a greater variety of products. Interestingly, despite the fact that there are more regular customers than new customers (almost 3x as much), new customers have purchased more products than regular customers.

Regular customer    15876776

Loyal customer      10284093

New customer         6243990


## 6. Creating a spending flag for each user based on the average price across all their orders

In [1]:
# using the following criteria:
# If the mean of the prices of products purchased by a user is lower than 10, then flag them as a “Low spender.”
# If the mean of the prices of products purchased by a user is higher than or equal to 10, then flag them as a “High spender.”

In [15]:
#Creating a column for the average price 
df_ords_prods_merged['avg_price'] = df_ords_prods_merged.groupby(['user_id'])['prices'].transform(np.mean)

In [16]:
# Creating another column for the spending flag
df_ords_prods_merged.loc[df_ords_prods_merged['avg_price'] >= 10, 'spending_flag'] = 'High spender'

In [17]:
df_ords_prods_merged.loc[df_ords_prods_merged['avg_price'] < 10, 'spending_flag'] = 'Low spender'

In [18]:
df_ords_prods_merged.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  ...  prices  _merge    price_range_loc     busiest_day  \
0          Soda  ...     9.0    both  Mid-range product  Regularly busy   
1          Soda  ...     9.0    both  Mid-range product  Regularly busy   
2          Soda  ...     9.0    both  Mid-range product  Regularly busy   
3          Soda  ...     9.0    both  Mid-range product      Least busy   
4          Soda  ...     9.0    both  Mid-range product      Least busy   
5          Soda  ...     9.0    both  Mid-range product  Regularly busy   
6          Soda  ...     9.0    both  Mid-range product  Regularly busy   
7          Soda  ...     9.0    both  Mid-range product  Regularly busy   
8          Soda  ...     9.0    both  Mid-range product  Regularly busy   
9          Soda  ...     9.0    both  Mid-range product      Least busy   
10         Soda  ...     9.0    both  Mid-range product  Regularly busy   
11         Soda  ...     9.0    both  Mid-range product  Regularly busy   
12         Soda  ...     9.0    both  Mid-range product  Regularly busy   
13         Soda  ...     9.0    both  Mid-range product  Regularly busy   
14         Soda  ...     9.0    both  Mid-range product  Regularly busy   

    Busiest_days busiest_period_of_day max_order      loyalty_flag  avg_price  \
0   Regular days           Most orders        10      New customer   6.367797   
1   Slowest days        Average orders        10      New customer   6.367797   
2   Slowest days           Most orders        10      New customer   6.367797   
3   Slowest days        Average orders        10      New customer   6.367797

In [19]:
#Checking the count of 'spending_flag' values for accuracy
df_ords_prods_merged['spending_flag'].value_counts(dropna = False)

Low spender     32285165
High spender      119694
Name: spending_flag, dtype: int64

In [20]:
df_ords_prods_merged.shape

(32404859, 22)

## 7. Creating an order frequency flag that marks the regularity of a user’s ordering behavior according to the median in the “days_since_prior_order” column to determine frequent versus non-frequent customers

In [2]:
# The criteria for the flag should be as follows:
# If the median of “days_since_prior_order” is higher than 20, then the customer should be labeled a “Non-frequent customer.”
# If the median is higher than 10 and lower than or equal to 20, then the customer should be labeled a “Regular customer.”
# If the median is lower than or equal to 10, then the customer should be labeled a “Frequent customer.”

In [22]:
# Creating a column for the median frequency
df_ords_prods_merged['median_frequency'] = df_ords_prods_merged.groupby(['user_id'])['days_since_prior_order'].transform(np.median)

In [26]:
# creating another column for the order_frequency flag
df_ords_prods_merged.loc[df_ords_prods_merged['median_frequency'] > 20, 'order_frequency_flag'] = 'Non-frequent customer'

In [27]:
df_ords_prods_merged.loc[(df_ords_prods_merged['median_frequency'] > 10) & (df_ords_prods_merged['median_frequency'] <= 20), 'order_frequency_flag'] = 'Regular customer'

In [28]:
df_ords_prods_merged.loc[df_ords_prods_merged['median_frequency'] <= 10, 'order_frequency_flag'] = 'Frequent customer'

In [29]:
df_ords_prods_merged.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  ...     busiest_day  Busiest_days  busiest_period_of_day  \
0          Soda  ...  Regularly busy  Regular days            Most orders   
1          Soda  ...  Regularly busy  Slowest days         Average orders   
2          Soda  ...  Regularly busy  Slowest days            Most orders   
3          Soda  ...      Least busy  Slowest days         Average orders   
4          Soda  ...      Least busy  Slowest days            Most orders   
5          Soda  ...  Regularly busy  Regular days         Average orders   
6          Soda  ...  Regularly busy  Busiest days            Most orders   
7          Soda  ...  Regularly busy  Busiest days            Most orders   
8          Soda  ...  Regularly busy  Busiest days            Most orders   
9          Soda  ...      Least busy  Slowest days            Most orders   
10         Soda  ...  Regularly busy  Busiest days            Most orders   
11         Soda  ...  Regularly busy  Regular days            Most orders   
12         Soda  ...  Regularly busy  Regular days            Most orders   
13         Soda  ...  Regularly busy  Busiest days            Most orders   
14         Soda  ...  Regularly busy  Busiest days            Most orders   

   max_order      loyalty_flag avg_price spending_flag median_freq  \
0         10      New customer  6.367797   Low spender        20.5   
1         10      New customer  6.367797   Low spender        20.5   
2         10      New customer  6.367797   Low spender        20.5   
3         10      New customer  6.367797   Low spender        20.5   
4         10      N

In [30]:
df_ords_prods_merged = df_ords_prods_merged.drop(columns = ['median_freq'])

In [31]:
df_ords_prods_merged.head(15)

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0    2539329        1             1                   2                  8   
1    2398795        1             2                   3                  7   
2     473747        1             3                   3                 12   
3    2254736        1             4                   4                  7   
4     431534        1             5                   4                 15   
5    3367565        1             6                   2                  7   
6     550135        1             7                   1                  9   
7    3108588        1             8                   1                 14   
8    2295261        1             9                   1                 16   
9    2550362        1            10                   4                  8   
10   2968173       15            15                   1                  9   
11   1870022       15            17                   2                 16   
12   1911383       15            18                   2                 11   
13   2715276       15            21                   1                  9   
14    487368       15            22                   1                 10   

    days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                      NaN         196                  1          0   
1                     15.0         196                  1          1   
2                     21.0         196                  1          1   
3                     29.0         196                  1          1   
4                     28.0         196                  1          1   
5                     19.0         196                  1          1   
6                     20.0         196                  1          1   
7                     14.0         196                  2          1   
8                      0.0         196                  4          1   
9                     30.0         196                  1          1   
10                     7.0         196                  2          0   
11                     8.0         196                  6          1   
12                     7.0         196                  1          1   
13                     7.0         196                  1          1   
14                    14.0         196                  1          1   

   product_name  ...    price_range_loc     busiest_day  Busiest_days  \
0          Soda  ...  Mid-range product  Regularly busy  Regular days   
1          Soda  ...  Mid-range product  Regularly busy  Slowest days   
2          Soda  ...  Mid-range product  Regularly busy  Slowest days   
3          Soda  ...  Mid-range product      Least busy  Slowest days   
4          Soda  ...  Mid-range product      Least busy  Slowest days   
5          Soda  ...  Mid-range product  Regularly busy  Regular days   
6          Soda  ...  Mid-range product  Regularly busy  Busiest days   
7          Soda  ...  Mid-range product  Regularly busy  Busiest days   
8          Soda  ...  Mid-range product  Regularly busy  Busiest days   
9          Soda  ...  Mid-range product      Least busy  Slowest days   
10         Soda  ...  Mid-range product  Regularly busy  Busiest days   
11         Soda  ...  Mid-range product  Regularly busy  Regular days   
12         Soda  ...  Mid-range product  Regularly busy  Regular days   
13         Soda  ...  Mid-range product  Regularly busy  Busiest days   
14         Soda  ...  Mid-range product  Regularly busy  Busiest days   

   busiest_period_of_day max_order      loyalty_flag avg_price spending_flag  \
0            Most orders        10      New customer  6.367797   Low spender   
1         Average orders        10      New customer  6.367797   Low spender   
2            Most orders        10      New customer  6.367797   Low spender   
3         Average orders        10      New customer  6.367797   Low spender   
4            Most orders        1

In [32]:
#Checking the count of 'order_frequency_flag' values
df_ords_prods_merged['order_frequency_flag'].value_counts(dropna = False)

Frequent customer        21559853
Regular customer          7208564
Non-frequent customer     3636437
NaN                             5
Name: order_frequency_flag, dtype: int64

In [33]:
df_ords_prods_merged.shape

(32404859, 24)

## 8. Exporting the dataframe as a pickle file

In [34]:
# Exporting 'df_ords_prods_merged_aggregated.pkl' to 'Prepared Data'
df_ords_prods_merged.to_pickle(os.path.join(path,'02 Data','Prepared Data','orders_products_merged_aggregated.pkl'))